internal/apex/ctlserver.go

```diff
@@ -0,0 +1,119 @@
+package apex
+
+import (
+	"fmt"
+	"net"
+	"net/rpc"
+	"net/rpc/jsonrpc"
+	"os"
+	"sync"
+)
+
+// TODO make this path configurable
+const UnixSocketPath = "/run/apex.sock"
+
+func (ax *Apex) CtlServerStart() error {
+	switch ax.os {
+	case Linux.String():
+		return ax.CtlServerStartLinux()
+
+	case Darwin.String():
+		ax.logger.Info("Ctl interface not yet supported on OSX")
+
+	case Windows.String():
+		ax.logger.Info("Ctl interface not yet supported on Windows")
+	}
+
+	return nil
+}
+
+type ApexCtl struct {
+	ax *Apex
+}
+
+func (ac *ApexCtl) Status(_ string, result *string) error {
+	var statusStr string
+	switch ac.ax.status {
+	case ApexStatusStarting:
+		statusStr = "Starting"
+	case ApexStatusAuth:
+		statusStr = "WaitingForAuth"
+	case ApexStatusRunning:
+		statusStr = "Running"
+	default:
+		statusStr = "Unknown"
+	}
+	res := fmt.Sprintf("Status: %s\n", statusStr)
+	if len(ac.ax.statusMsg) > 0 {
+		res += ac.ax.statusMsg
+	}
+	*result = res
+	return nil
+}
+
+func (ac *ApexCtl) Version(_ string, result *string) error {
+	*result = ac.ax.version
+	return nil
+}
+
+func (ax *Apex) CtlServerStartLinux() error {
+	os.Remove(UnixSocketPath)
+	l, err := net.ListenUnix("unix", &net.UnixAddr{Name: UnixSocketPath, Net: "unix"})
+	if err != nil {
+		ax.logger.Error("Error creating unix socket: ", err)
+		return err
+	}
+	ax.wg.Add(1)
+	go func() {
+		defer l.Close()
+		defer ax.wg.Done()
+
+		ac := new(ApexCtl)
+		ac.ax = ax
+		err := rpc.Register(ac)
+		if err != nil {
+			ax.logger.Errorln("Error on rpc.Register(): ", err)
+		}
+
+		// We accept connections and send them back over this channel
+		// so we can select waiting on new connections and a message
+		// to stop the CtlServer at the same time.
+		connChan := make(chan net.Conn)
+		go func() {
+			for {
+				conn, err := l.Accept()
+				if err != nil {
+					// TODO - if Accept() returns an error because the listener
+					// has been closed, we should just silently end this goroutine.
```

Comment:
@dave-tucker I didn't see a clean way to do this. Any ideas?

Reply:
Yeah from the previous round of comments I think I'd mentioned keeping this in a `RunCtlServer` function which was a blocking function. i.e
```go
func runCtlServer(wg *sync.Waitgroup, stopCh <-chan struct{}) error {}
```

The logic would be roughly:

1. Have a `for { select {} }` that has 2 cases: `case <-ax.ctlServerStop:` which waits will run when we're shutting down and `default` which is run when no other cases are available i.e we're not shutting down.
1. Any error in `conn, err := l.Accept()` can be returned directly i.e `return err`
2. If no error from `Accept()`, increment the waitgroup by 1 and spawn a goroutine to service the connection - that goroutine has a `defer wg.Done()`


That way, we can handle any errors that happen properly... We might want to try and run the server again for example.

Reply:
I pushed another iteration, but it's not quite what you described here. I do have it properly catching errors and restarting the control interface routine if an error occurs. Take a look at your leisure and then let's jump on video chat to talk through feedback.

Thanks again for the detailed reviews!

Reply:
I also left the rudimentary state tracking. I figured we can replace it in a follow up PR.